conjure-undertow-lib/src/main/java/com/palantir/conjure/java/undertow/lib/Parameters.java

```diff
@@ -0,0 +1,79 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.undertow.lib;
+
+import com.google.common.collect.Multimap;
+import com.palantir.logsafe.SafeArg;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.util.AttachmentKey;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Utility class to put different types of safe or unsafe parameters in a multimap attached the exchange.
+ * It requires the exchanged o have been previously through a ParametersHandler which would have created the attached
+ * multimaps..
+ */
+public final class Parameters {
+
+    private static final Logger log = LoggerFactory.getLogger(Parameters.class);
+
+    public static final AttachmentKey<Multimap<String, Object>> SAFE_PARAMS_ATTACH_KEY =
+            AttachmentKey.create(Multimap.class);
+    public static final AttachmentKey<Multimap<String, Object>> UNSAFE_PARAMS_ATTACH_KEY =
+            AttachmentKey.create(Multimap.class);
+
+    private Parameters() {}
+
+    public static void putSafePathParam(HttpServerExchange exchange, String key, Object value) {
+        putParam(exchange, SAFE_PARAMS_ATTACH_KEY, key, value);
+    }
+
+    public static void putUnsafePathParam(HttpServerExchange exchange, String key, Object value) {
+        putParam(exchange, UNSAFE_PARAMS_ATTACH_KEY, key, value);
+    }
+
+    public static void putSafeQueryParam(HttpServerExchange exchange, String key, Object value) {
+        putParam(exchange, SAFE_PARAMS_ATTACH_KEY, key, value);
+    }
+
+    public static void putUnsafeQueryParam(HttpServerExchange exchange, String key, Object value) {
+        putParam(exchange, UNSAFE_PARAMS_ATTACH_KEY, key, value);
+    }
+
+    public static void putSafeHeaderParam(HttpServerExchange exchange, String key, Object value) {
+        putParam(exchange, SAFE_PARAMS_ATTACH_KEY, key, value);
+    }
+
+    public static void putUnsafeHeaderParam(HttpServerExchange exchange, String key, Object value) {
+        putParam(exchange, UNSAFE_PARAMS_ATTACH_KEY, key, value);
+    }
+
+    private static void putParam(
+            HttpServerExchange exchange,
+            AttachmentKey<Multimap<String, Object>> attachmentKey,
+            String key,
+            Object value) {
+        Optional<Multimap<String, Object>> multimap = Optional.ofNullable(exchange.getAttachment(attachmentKey));
+        if (multimap.isPresent()) {
+            multimap.get().put(key, value);
+        } else {
+            log.warn("Attachment key not found. Impossible to put param {}.", SafeArg.of("param-name", key));
```

Comment:
nit:
```

log.warn("Failed to attached endpoint parameter because the the required exchange attachment does not exist. Did this request get handled by the ParametersHandler?", SafeArg.of("attachmentKey", attachmentKey), SafeArg.of("parameterName", key));
```